src/server/pkg/transactionenv/env.go

```diff
@@ -0,0 +1,373 @@
+// either directly run the request through Auth or append it to the active
+// transaction, depending on if there is an active transaction in the client
+// context.
+type AuthWrites interface {
+	SetScope(*auth.SetScopeRequest) (*auth.SetScopeResponse, error)
+	SetACL(*auth.SetACLRequest) (*auth.SetACLResponse, error)
+}
+
+// PfsTransactionDefer is the interface that PFS provides for deferring certain
+// tasks until the end of a transaction.  It is defined here to avoid a circular
+// dependency.
+type PfsTransactionDefer interface {
+	PropagateCommit(branch *pfs.Branch) error
+	Run() error
+}
+
+// TransactionContext is a helper type to encapsulate the state for a given
+// set of operations being performed in the Pachyderm API.  When a new
+// transaction is started, a context will be created for it containing these
+// objects, which will be threaded through to every API call:
+//   ctx: the client context which initiated the operations being performed
+//   pachClient: the APIClient associated with the client context ctx
+//   stm: the object that controls transactionality with etcd.  This is to ensure
+//     that all reads and writes are consistent until changes are committed.
+//   txnEnv: a struct containing references to each API server, it can be used
+//     to make calls to other API servers (e.g. checking auth permissions)
+//   pfsDefer: an interface for ensuring certain PFS cleanup tasks are performed
+//     properly (and deduped) at the end of the transaction.
+type TransactionContext struct {
+	ctx        context.Context
+	pachClient *client.APIClient
+	stm        col.STM
+	txnEnv     *TransactionEnv
+	pfsDefer   PfsTransactionDefer
+}
+
+// Auth returns a reference to the Auth API Server so that transactionally-
+// supported methods can be called across the API boundary without using RPCs
+// (which will not maintain transactional guarantees)
+func (t *TransactionContext) Auth() AuthTransactionServer {
+	return t.txnEnv.authServer
+}
+
+// Pfs returns a reference to the PFS API Server so that transactionally-
+// supported methods can be called across the API boundary without using RPCs
+// (which will not maintain transactional guarantees)
+func (t *TransactionContext) Pfs() PfsTransactionServer {
+	return t.txnEnv.pfsServer
+}
+
+// Client returns an APIClient object for making downstream requests to API
+// servers
+func (t *TransactionContext) Client() *client.APIClient {
+	return t.pachClient
+}
+
+// ClientContext returns the client context object associated with the
+// client from Client()
+func (t *TransactionContext) ClientContext() context.Context {
+	return t.ctx
+}
+
+// Stm returns the STM object for transactionality with etcd
+func (t *TransactionContext) Stm() col.STM {
+	return t.stm
+}
+
+// PfsDefer returns a reference to the object for deferring PFS cleanup tasks to
+// the end of the transaction
+func (t *TransactionContext) PfsDefer() PfsTransactionDefer {
+	return t.pfsDefer
+}
```

Comment:
Might as well make `pfsDefer` public if we're going to have an accessor method.

Reply:
Done (as well as the other getters in this struct)